README.rst
==============================================================================
ftw.flamegraph
==============================================================================

This package provides a statistical profiler for Plone and creates flame graphs
for visualization of profiled code. It allows you to profile a single request
in Plone.

Because of it's low overhead, it can also be used for profiling Plone sites in
production.


Introduction
------------

In contrast to deterministic profilers that capture all code paths, ftw.flamegraph
samples the call stack periodically (by default 1000 times per second) which
reveals where time is being spent and involves less overhead.

Flame graphs are a way to visualize statistical profile data, allowing hot code-paths
to be identified quickly. For more information about flame graphs see:
http://www.brendangregg.com/flamegraphs.html


Installation
------------

Install ftw.flamegraph by adding it to your buildout::

    [buildout]
    ...
    eggs =
        ftw.flamegraph


and then running ``bin/buildout``


Usage
-----

To profile a request simply add the parameter ``flamegraph=1`` to the URL.

Example::

 http://localhost:8080/Plone?flamegraph=1

This will profile the request and render a SVG containing a flame graph.

You can customize the sampling interval by providing the ``interval`` parameter.

Example::

 http://localhost:8080/Plone?flamegraph=1&interval=0.0005

This will sample the call stack 2000 times per second. Too small intervals will
create more overhead while too large intervals may lead to missing hot code paths.


Use in tests
------------

The flamegraph decorator can be used for wrapping tests (or other functions) and
producing a flamegraph:

.. code:: python

    from ftw.flamegraph import flamegraph

    @flamegraph()
    def test():
        do_things()


Options:

- ``open_svg``: Open the SVG with the systems "open" command (default: ``True``)
- ``interval``: The interval for taking snapshots (default: ``0.001``)



Implementation notes
--------------------

ftw.flamegraph sets an interval timer which decrements in real time (ITIMER_REAL)
and sends a signal (SIGALRM) upon expiration. Every time the signal gets delivered,
the current thread's call stack gets recorded.

While timers that decrement only when the process is executing (ITIMER_PROF and
ITIMER_VIRTUAL) might be more appropriate for profiling, this is not possible with
Zope 2. In Python those timers only decrement when the main thread is
executing, but in Zope 2 the main thread is an asyncore loop that waits for I/O
with a timeout of 30 seconds, making it inappropriate for periodic sampling.

To get meaningful results you should make sure there are no other CPU intensive
processes executing on the same CPU as the profiled process.


Links
-----

- Issue Tracker: https://github.com/4teamwork/ftw.flamegraph/issues
- Source Code: https://github.com/4teamwork/ftw.flamegraph


Copyright
---------

This package is copyright by `4teamwork <http://www.4teamwork.ch/>`_.

``ftw.flamegraph`` is licensed under GNU General Public License, version 2.


CHANGES.rst
Changelog
=========


1.1.1 (unreleased)
------------------

- Add collective.indexing / CMFPlone.indexing support. [busykoala]
- Set status code to prevent redirect and the loss of the flamegraph. [busykoala]


1.1.0 (2017-07-18)
------------------

- Add decorator for generating flamegraphs. [jone]


1.0.1 (2016-12-22)
------------------

- Fix manifest.in.


1.0.0 (2016-10-04)
------------------

- Initial release.
  [buchi]


ftw/flamegraph/tests/test_sampler.py
from unittest import TestCase
from ftw.flamegraph.testing import FLAMEGRAPH_INTEGRATION_TESTING
from ftw.flamegraph.sampler import Sampler
import time


class TestInstall(TestCase):
    layer = FLAMEGRAPH_INTEGRATION_TESTING

    def test_sampler_initalization(self):
        from ftw.flamegraph import sampler
        self.assertIsNotNone(sampler)


class TestSampler(TestCase):

    def test_sampling(self):
        def function(end):
            while time.time() < end:
                pass
        sampler = Sampler()
        sampler.start()
        function(time.time() + 0.01)
        sampler.stop()
        self.assertGreater(len(sampler.stacks), 0)
        self.assertEqual('function', sampler.stacks[0][0][2])

    def test_folded_stacks_output(self):
        sampler = Sampler()
        sampler.stacks = [[
            ('/path/of/module.py', 12, 'methodname'),
            ('/path/of/module.py', 50, 'other_method'),
            ('/path/of/other_module.py', 3, 'methodname'),
        ]]
        self.assertEqual(
            'methodname (/path/of/other_module.py:3);'
            'other_method (/path/of/module.py:50);'
            'methodname (/path/of/module.py:12) 1\n', sampler.folded_stacks())


ftw/flamegraph/tests/test_flamegraph.py
from unittest import TestCase
from ftw.flamegraph.flamegraph import FlameGraph


class TestFlameGraph(TestCase):

    def test_1(self):
        stacks = [
            [
                ('/path/of/other_module.py', 3, 'methodname'),
                ('/path/of/module.py', 50, 'other_method'),
                ('/path/of/module.py', 12, 'methodname'),
            ],
            [
                ('/path/of/module.py', 50, 'other_method'),
                ('/path/of/module.py', 12, 'methodname'),
            ],
            [
                ('/path/of/other_module.py', 100, 'other_method'),
                ('/path/of/module.py', 12, 'methodname'),
            ],
        ]
        fg = FlameGraph(stacks)


ftw/flamegraph/sampler.py
import signal
import sys
import thread
import time


class Sampler(object):
    """A call stack sampler
    """
    timer = signal.ITIMER_REAL
    signal = signal.SIGALRM

    def __init__(self, interval=0.001):
        self.interval = interval
        signal.signal(self.signal, self.handler)
        signal.siginterrupt(self.signal, False)
        self.reset()

    def reset(self, interval=None):
        self.stacks = []
        self.stopping = False
        self.stopped = False
        self.samples_taken = 0
        self.sample_time = 0

        if interval is not None:
            self.interval = interval

    def start(self):
        self.thread_id = thread.get_ident()
        self.stopping = False
        self.stopped = False
        signal.setitimer(self.timer, self.interval, self.interval)

    def stop(self):
        self.stopping = True
        self.wait()

    def wait(self):
        while not self.stopped:
            time.sleep(self.interval)

    def handler(self, sig, current_frame):
        start = time.time()
        if self.stopping:
            signal.setitimer(self.timer, 0, 0)
            self.stopped = True
            return
        current_tid = thread.get_ident()
        for tid, frame in sys._current_frames().iteritems():
            # Only collect samples for our thread
            if tid != self.thread_id:
                continue
            if tid == current_tid:
                frame = current_frame
            frames = []
            while frame is not None:
                code = frame.f_code
                frames.append((code.co_filename, frame.f_lineno, code.co_name))
                frame = frame.f_back
            self.stacks.append(frames)
        end = time.time()
        self.samples_taken += 1
        self.sample_time += (end - start)

    def folded_stacks(self):
        """Fold stack samples into single lines for flame graph generation"""
        output = ""
        previous = None
        previous_count = 1
        for stack in self.stacks:
            funcs = map("{0[2]} ({0[0]}:{0[1]})".format, reversed(stack))
            current = ";".join(funcs)
            if current == previous:
                previous_count += 1
            else:
                if previous is not None:
                    output += "%s %d\n" % (previous, previous_count)
                previous_count = 1
                previous = current
        output += "%s %d\n" % (previous, previous_count)
        return output


ftw/flamegraph/__init__.py
# -*- coding: utf-8 -*-
from StringIO import StringIO
from ZPublisher import Publish
from contextlib import contextmanager
from ftw.flamegraph.sampler import Sampler
from functools import wraps
from urlparse import parse_qs

import logging
import os
import pkg_resources
import shutil
import sys
import tempfile
import time


IS_PLONE_5 = pkg_resources.get_distribution('Products.CMFPlone').version >= '5'


if IS_PLONE_5:
    from Products.CMFCore.indexing import processQueue
else:
    # optional collective.indexing support
    try:
        from collective.indexing.queue import processQueue
    except ImportError:
        def processQueue():
            pass


sampler = None
logger = logging.getLogger("ftw.flamegraph")
flamegraph_script = os.path.join(os.path.dirname(__file__), 'flamegraph.pl')


def flamegraph(open_svg=False, open_command='open {}', interval=0.001):
    """The @flamegraph() decorator wraps a function and produces a flamegraph
    from calls within this function.
    """
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with profile(interval, print_summary=True) as sampler:
                return_value = func(*args, **kwargs)
                processQueue()

            svg_file_path = make_svg(sampler)
            print '@flamegraph(', svg_file_path, ')'
            if open_svg:
                os.system(open_command.format(svg_file_path))

            return return_value

        return wrapper
    return decorator


def publish_module_standard(
    module_name, stdin=sys.stdin, stdout=sys.stdout, stderr=sys.stderr,
        environ=os.environ, debug=0, request=None, response=None):

    if 'flamegraph=1' in request.get('QUERY_STRING', ''):
        qs = parse_qs(request['QUERY_STRING'])
        interval = qs.get('interval', ['0.001'])
        try:
            interval = float(interval[0])
        except ValueError:
            interval = 0.001

        response_stdout = response.stdout
        response.stdout = StringIO()

        with profile(interval):
            ret = publish_module_standard.original(
                module_name, stdin=stdin, stdout=stdout, stderr=stderr,
                environ=environ, debug=debug, request=request,
                response=response)
            processQueue()

        response.stdout = response_stdout
        # do not redirect (otherwise the flamegraph is lost)
        response.setStatus(200)

        svg_file_path = make_svg(sampler)
        with open(svg_file_path, 'rb') as f:
            response.setBody(f.read())
        response.setHeader('Content-Type', 'image/svg+xml')
        response.stdout.write(str(response))

        shutil.rmtree(os.path.dirname(svg_file_path))
        return ret

    else:
        return publish_module_standard.original(
            module_name, stdin=stdin, stdout=stdout, stderr=stderr,
            environ=environ, debug=debug, request=request,
            response=response)

publish_module_standard.original = Publish.publish_module_standard
Publish.publish_module_standard = publish_module_standard


def make_svg(sampler):
    tempdir = tempfile.mkdtemp()
    flame_file_path = os.path.join(tempdir, 'profile.flame')
    svg_file_path = os.path.join(tempdir, 'profile.svg')

    with open(flame_file_path, "wb") as f:
        f.write(sampler.folded_stacks().encode('utf-8'))

    os.system(
        '%(flamegraph_script)s --egghash --color rainbow %(flame_file)s > '
        '%(svg_file)s' % (dict(
            flamegraph_script=flamegraph_script,
            flame_file=flame_file_path,
            svg_file=svg_file_path)))

    return svg_file_path


@contextmanager
def profile(interval, print_summary=False):
    global sampler
    if sampler is None:
        sampler = Sampler()

    sampler.reset(interval=interval)
    sampler.start()
    start = time.time()
    yield sampler
    end = time.time()
    sampler.stop()

    message = "Samples taken: %s, sample time: %s ms, duration: %s ms" % (
        sampler.samples_taken,
        sampler.sample_time * 1000,
        (end - start) * 1000
    )
    logger.info(message)
    if print_summary:
        print '\n', message


def initialize(context):
    # The sampler needs to be initialized on the main thread
    global sampler
    sampler = Sampler()


ftw/flamegraph/platform.py
import signal

if hasattr(signal, 'setitimer'):
    from signal import setitimer, ITIMER_REAL, ITIMER_VIRTUAL, ITIMER_PROF
else:
    # python2.5 doesn't have setitimer
    import ctypes.util
    libc = ctypes.CDLL(ctypes.util.find_library("c"))

    class Timeval(ctypes.Structure):
        _fields_ = [('tv_sec', ctypes.c_long), ('tv_usec', ctypes.c_long)]

    class Itimerval(ctypes.Structure):
        _fields_ = [('it_interval', Timeval), ('it_value', Timeval)]
    libc.setitimer.argtypes = [ctypes.c_int, ctypes.POINTER(Itimerval),
                               ctypes.POINTER(Itimerval)]

    def seconds_to_timeval(seconds):
        return Timeval(int(seconds), int((seconds % 1) * 1000000))

    def setitimer(which, seconds, interval):
        libc.setitimer(which,
                       Itimerval(seconds_to_timeval(interval),
                                 seconds_to_timeval(seconds)),
                       # We should get the output value here to return,
                       # but we don't need it right now.
                       None)
    ITIMER_REAL = 0
    ITIMER_VIRTUAL = 1
    ITIMER_PROF = 2


ftw/flamegraph/flamegraph.py
import xml.etree.ElementTree as etree


class FlameGraph(object):

    def __init__(self, stacks):
        self.boxes = {}  # (frame, depth, endtime) -> starttime
        self.depth_samples = {}  # depth -> number of samples
        self.stacks = {}
        self.stack_counts = {}
        self.maxdepth = 0

        for stack in stacks:
            frame = map("{0[2]} ({0[0]}:{0[1]})".format, reversed(stack))
            key = ','.join(frame)
            if key not in self.stacks:
                self.stacks[key] = frame
                self.stack_counts[key] = 0
            self.stack_counts[key] += 1

        self.calculate_boxes()

    def calculate_boxes(self):
        # import pdb; pdb.set_trace()
        previous_stack = []
        starts = {}  # (frame, depth) -> starttime
        time = 0

        for key in sorted(self.stacks.keys()):
            frames = self.stacks[key]
            count = self.stack_counts[key]

            # Prepend an empty frame to every stack to represent the "all
            # samples" frame
            frames.insert(0, '')

            if len(frames) - 1 > self.maxdepth:
                self.maxdepth = len(frames) - 1

            # Determine the number of same frames of the current and the
            # previous stack
            same_frames = 0
            for i, frame in enumerate(previous_stack):
                if i > len(frames):
                    break
                if frame == frames[i]:
                    same_frames += 1
                else:
                    break

            # Frames ending here
            for i in reversed(range(same_frames, len(previous_stack))):
                starts_key = (previous_stack[i], i)
                ends_key = (previous_stack[i], i, time)
                self.boxes[ends_key] = starts[starts_key]
                self.depth_samples[i] += (time - starts[starts_key])
                del starts[starts_key]

            # Mark frames not present in previous stack as starting here
            for i in range(same_frames, len(frames)):
                starts[(frames[i], i)] = time

                if i not in self.depth_samples:
                    self.depth_samples[i] = 0

            time += count
            previous_stack = frames

        # Frames ending at the end
        for i in reversed(range(0, len(previous_stack))):
            starts_key = (previous_stack[i], i)
            ends_key = (previous_stack[i], i, time)
            self.boxes[ends_key] = starts[starts_key]
            self.depth_samples[i] += (time - starts[starts_key])

        self.total_time = time

    def draw_svg(self):
        width = 1200
        width_per_sample = width / self.total_time
        frame_height = 16
        height = self.maxdepth * frame_height
        root = etree.Element(
            'svg',
            version='1.1',
            width=str(width),
            height=str(height),
            viewBox='0 0 {} {}'.format(width, height),
            xmlns='http://www.w3.org/2000/svg',
        )
        for (frame, depth, end_time), start_time in self.boxes.items():
            root.append(etree.Element(
                'rect',
                 x=str(start_time * width_per_sample),
                 y=str(height - (depth + 1) * frame_height + 1),
                 width=str((end_time - start_time) * width_per_sample),
                 fill="rgb(224,0,2)",
                 rx='2',
                 ry='2',
            ))

        return etree.tostring(root, encoding='UTF-8')


ftw/flamegraph/collector.py
from __future__ import with_statement
from ftw.flamegraph import platform
from six.moves import _thread

import collections
import signal
import six
import sys
import time


class Collector(object):
    MODES = {
        'prof': (platform.ITIMER_PROF, signal.SIGPROF),
        'virtual': (platform.ITIMER_VIRTUAL, signal.SIGVTALRM),
        'real': (platform.ITIMER_REAL, signal.SIGALRM),
    }

    def __init__(self, interval=0.01, mode='virtual'):
        self.interval = interval
        self.mode = mode
        assert mode in Collector.MODES
        timer, sig = Collector.MODES[self.mode]
        signal.signal(sig, self.handler)
        signal.siginterrupt(sig, False)
        self.reset()

    def reset(self):
        self.stacks = list()
        self.samples_remaining = 0
        self.stopping = False
        self.stopped = False

        self.samples_taken = 0
        self.sample_time = 0

    def start(self, duration=30.0):
        self.thread_id = _thread.get_ident()
        self.stopping = False
        self.stopped = False
        self.samples_remaining = int(duration / self.interval)
        timer, sig = Collector.MODES[self.mode]
        platform.setitimer(timer, self.interval, self.interval)

    def stop(self):
        self.stopping = True
        self.wait()

    def wait(self):
        while not self.stopped:
            pass  # need busy wait; ITIMER_PROF doesn't proceed while sleeping

    def handler(self, sig, current_frame):
        start = time.time()
        self.samples_remaining -= 1
        if self.samples_remaining <= 0 or self.stopping:
            platform.setitimer(Collector.MODES[self.mode][0], 0, 0)
            self.stopped = True
            return
        current_tid = _thread.get_ident()
        for tid, frame in six.iteritems(sys._current_frames()):
            if tid != self.thread_id:
                continue
            if tid == current_tid:
                frame = current_frame
            frames = []
            while frame is not None:
                code = frame.f_code
                frames.append((code.co_filename, frame.f_lineno, code.co_name))
                frame = frame.f_back
            self.stacks.append(frames)
        end = time.time()
        self.samples_taken += 1
        self.sample_time += (end - start)


class CollectorFormatter(object):
    """
    Abstract class for output formats
    """
    def format(self, collector):
        raise Exception("not implemented")

    def store(self, collector, filename):
        with open(filename, "wb") as f:
            f.write(self.format(collector).encode('utf-8'))


class PlopFormatter(CollectorFormatter):
    """
    Formats stack frames for plop.viewer
    """
    def __init__(self, max_stacks=50):
        self.max_stacks = 50

    def format(self, collector):
        # defaultdict instead of counter for pre-2.7 compatibility
        stack_counts = collections.defaultdict(int)
        for frames in collector.stacks:
            stack_counts[tuple(frames)] += 1
        stack_counts = dict(sorted(six.iteritems(stack_counts),
                                   key=lambda kv: -kv[1])[:self.max_stacks])
        return repr(stack_counts)


class FlamegraphFormatter(CollectorFormatter):
    """
    Creates Flamegraph files
    """
    def format(self, collector):
        output = ""
        previous = None
        previous_count = 1
        for stack in collector.stacks:
            current = self.format_flame(stack)
            if current == previous:
                previous_count += 1
            else:
                output += "%s %d\n" % (previous, previous_count)
                previous_count = 1
                previous = current
        output += "%s %d\n" % (previous, previous_count)
        return output

    def format_flame(self, stack):
        funcs = map("{0[2]} ({0[0]}:{0[1]})".format, reversed(stack))
        return ";".join(funcs)
